examples/hyper_server.rs

```diff
@@ -0,0 +1,161 @@
+	}
+
+	// Just a little ceremony we need to go through to return the correct response key:
+	let mut accept_key_buf = [0; 32];
+	let accept_key = generate_websocket_accept_key(key.as_bytes(), &mut accept_key_buf);
+
+	let response = Response::builder()
+		.status(hyper::StatusCode::SWITCHING_PROTOCOLS)
+		.header(hyper::header::CONNECTION, "upgrade")
+		.header(hyper::header::UPGRADE, "websocket")
+		.header("Sec-WebSocket-Accept", accept_key)
+		.body(Body::empty())
+		.expect("bug: failed to build response");
+
+	Ok((response, hyper::upgrade::on(req)))
+}
+
+/// Echo any messages we get from the client back to them
+async fn websocket_echo_messages(on_upgrade: hyper::upgrade::OnUpgrade) -> Result<(), BoxedError> {
+	// Wait for the request to upgrade, and pass the stream we get back to Soketto to handle the WS connection:
+	let stream = on_upgrade.await?;
+	let server = handshake::Server::new(BufReader::new(BufWriter::new(stream.compat())));
+	let (mut sender, mut receiver) = server.into_builder().finish();
+
+	// Echo any received messages back to the client:
+	let mut message = Vec::new();
+	loop {
+		message.clear();
+		match receiver.receive_data(&mut message).await {
+			Ok(soketto::Data::Binary(n)) => {
+				assert_eq!(n, message.len());
+				sender.send_binary_mut(&mut message).await?;
+				sender.flush().await?
+			}
+			Ok(soketto::Data::Text(n)) => {
+				assert_eq!(n, message.len());
+				if let Ok(txt) = std::str::from_utf8(&message) {
+					sender.send_text(txt).await?;
+					sender.flush().await?
+				} else {
+					break;
+				}
+			}
+			Err(soketto::connection::Error::Closed) => break,
+			Err(e) => {
+				eprintln!("Websocket connection error: {}", e);
+				break;
+			}
+		}
+	}
+
+	Ok(())
+}
+
+/// Defined in RFC 6455. this is how we convert the Sec-WebSocket-Key in a request into a
+/// Sec-WebSocket-Accept that we return in the response.
+fn generate_websocket_accept_key<'a>(key: &[u8], buf: &'a mut [u8; 32]) -> &'a [u8] {
+	// Defined in RFC 6455, we append this to the key to generate the response:
+	const KEY: &[u8] = b"258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
+
+	use sha1::{Digest, Sha1};
+	let mut digest = Sha1::new();
+	digest.update(key);
+	digest.update(KEY);
+	let d = digest.finalize();
+
+	let n = base64::encode_config_slice(&d, base64::STANDARD, buf);
+	&buf[..n]
+}
+
+/// Check if a request is a websocket upgrade request.
+pub fn is_upgrade_request<B>(request: &hyper::Request<B>) -> bool {
+	header_contains_value(request.headers(), hyper::header::CONNECTION, b"upgrade")
+		&& header_contains_value(request.headers(), hyper::header::UPGRADE, b"websocket")
+}
+
+/// Check if there is a header of the given name containing the wanted value.
+fn header_contains_value(headers: &hyper::HeaderMap, header: hyper::header::HeaderName, value: &[u8]) -> bool {
+	pub fn trim(x: &[u8]) -> &[u8] {
+		let from = match x.iter().position(|x| !x.is_ascii_whitespace()) {
+			Some(i) => i,
+			None => return &x[0..0],
```

Comment:
I suppose you could just return an empty slice for readability....

Reply:
That's a very good point! I probably assumed I needed to preserve the lifetime when I ported this code over, but of course it'll happily be static :)